// diagram-library-angular/esm5/lib/node-list-view.component.js
import { __decorate } from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import * as MindFusion from 'diagram-library';
var NodeListView = /** @class */ (function () {
    function NodeListView() {
        this.enabled = true;
        this.icon_size = new MindFusion.Drawing.Size(24, 24);
        this.node_size = new MindFusion.Drawing.Size(24, 24);
        this.padding = 5;
        this.orientation = MindFusion.Diagramming.Orientation.Vertical;
        this.tooltipDelay = 1000;
        this.shapeLibraryLocation = "";
    }
    NodeListView.prototype.ngAfterViewInit = function () {
        if (this.node_id === undefined)
            this.node_id = "nodeList";
        this.canvas.nativeElement.id = this.node_id;
        this.nodeList = MindFusion.Diagramming.NodeListView.create(document.getElementById(this.node_id));
        this.nodeList.setEnabled(this.enabled);
        this.nodeList.setIconSize(this.icon_size);
        this.nodeList.setDefaultNodeSize(this.node_size);
        this.nodeList.setPadding(this.padding);
        this.nodeList.setOrientation(this.orientation);
        this.nodeList.setShapeLibraryLocation(this.shapeLibraryLocation);
        this.nodeList.setTooltipDelay(this.tooltipDelay);
        this.addNode(this.nodes, this.caption);
    };
    NodeListView.prototype.addNode = function (nodes, caption) {
        for (var i = 0; i < nodes.length; ++i) {
            this.nodeList.addNode(nodes[i], caption[i]);
        }
    };
    __decorate([
        ViewChild('nodeList', { static: true })
    ], NodeListView.prototype, "canvas", void 0);
    __decorate([
        Input('id')
    ], NodeListView.prototype, "node_id", void 0);
    __decorate([
        Input('enabled')
    ], NodeListView.prototype, "enabled", void 0);
    __decorate([
        Input('iconSize')
    ], NodeListView.prototype, "icon_size", void 0);
    __decorate([
        Input('defaultNodeSize')
    ], NodeListView.prototype, "node_size", void 0);
    __decorate([
        Input('padding')
    ], NodeListView.prototype, "padding", void 0);
    __decorate([
        Input('orientation')
    ], NodeListView.prototype, "orientation", void 0);
    __decorate([
        Input('tooltipDelay')
    ], NodeListView.prototype, "tooltipDelay", void 0);
    __decorate([
        Input('shapeLibraryLocation')
    ], NodeListView.prototype, "shapeLibraryLocation", void 0);
    __decorate([
        Input('captions')
    ], NodeListView.prototype, "caption", void 0);
    __decorate([
        Input('nodes')
    ], NodeListView.prototype, "nodes", void 0);
    NodeListView = __decorate([
        Component({
            selector: 'node-list-view',
            template: "<div style=\"overflow-y: auto; overflow-x: hidden; width: 230px; height: 200px;\">\n                 <canvas #nodeList width=\"230\" height=\"200\" ></canvas>\n               </div>"
        })
    ], NodeListView);
    return NodeListView;
}());
export { NodeListView };
