// diagram-library-angular/esm5/lib/ruler.component.js
import { __decorate } from "tslib";
import { Component, ContentChild, Input, ViewChild } from '@angular/core';
import * as MindFusion from 'diagram-library';
import { DiagramView } from './diagram-library-angular.component';
var Ruler = /** @class */ (function () {
    function Ruler() {
        this.gr_unit = MindFusion.Drawing.GraphicsUnit.Millimeter;
        this.en_guides = true;
        this.vert_visible = true;
        this.hor_visible = true;
        this.pointer_color = "red";
        this.text_color = "#000000";
        this.guide_color = "#000000";
        this.projection_color = "lightSteelBlue";
        this.back_color = "#dddddd";
        this.fore_Color = "#000000";
    }
    Ruler.prototype.ngAfterContentInit = function () {
        if (this.ruler_id === undefined)
            this.ruler_id = "ruler";
        this.div_id.nativeElement.id = this.ruler_id;
        var ruler = new MindFusion.Diagramming.Ruler(document.getElementById(this.ruler_id));
        if (this.input && (this.input instanceof DiagramView)) {
            ruler.setDiagram(this.input.diagram);
        }
        else {
            console.error('ruler requirs diagram-view');
        }
        ruler.setUnit(this.gr_unit);
        ruler.setEnableGuides(this.en_guides);
        ruler.setProjectRotatedBounds(this.pr_bounds);
        ruler.setNegatedX(this.neg_x);
        ruler.setNegatedY(this.neg_y);
        ruler.setVerticalScaleVisible(this.vert_visible);
        ruler.setHorizontalScaleVisible(this.hor_visible);
        ruler.setPointerColor(this.pointer_color);
        ruler.setTextColor(this.text_color);
        ruler.setGuideColor(this.guide_color);
        ruler.setProjectionColor(this.projection_color);
        ruler.setBackColor(this.back_color);
        ruler.setForeColor(this.fore_Color);
    };
    __decorate([
        ContentChild(DiagramView)
    ], Ruler.prototype, "input", void 0);
    __decorate([
        ViewChild('id', { static: true })
    ], Ruler.prototype, "div_id", void 0);
    __decorate([
        Input('unit')
    ], Ruler.prototype, "gr_unit", void 0);
    __decorate([
        Input('enableGuides')
    ], Ruler.prototype, "en_guides", void 0);
    __decorate([
        Input('projectRotatedBounds')
    ], Ruler.prototype, "pr_bounds", void 0);
    __decorate([
        Input('negatedY')
    ], Ruler.prototype, "neg_y", void 0);
    __decorate([
        Input('negatedX')
    ], Ruler.prototype, "neg_x", void 0);
    __decorate([
        Input('verticalScaleVisible')
    ], Ruler.prototype, "vert_visible", void 0);
    __decorate([
        Input('horizontalScaleVisible')
    ], Ruler.prototype, "hor_visible", void 0);
    __decorate([
        Input('id')
    ], Ruler.prototype, "ruler_id", void 0);
    __decorate([
        Input('pointerColor')
    ], Ruler.prototype, "pointer_color", void 0);
    __decorate([
        Input('textColor')
    ], Ruler.prototype, "text_color", void 0);
    __decorate([
        Input('guideColor')
    ], Ruler.prototype, "guide_color", void 0);
    __decorate([
        Input('projectionColor')
    ], Ruler.prototype, "projection_color", void 0);
    __decorate([
        Input('backColor')
    ], Ruler.prototype, "back_color", void 0);
    __decorate([
        Input('foreColor')
    ], Ruler.prototype, "fore_Color", void 0);
    Ruler = __decorate([
        Component({
            selector: 'ruler',
            template: "<div #id>\n                   <ng-content></ng-content>\n               </div>"
        })
    ], Ruler);
    return Ruler;
}());
export { Ruler };
