// diagram-library-angular/esm2015/lib/overview.component.js
import { __decorate } from "tslib";
import { Component, ViewChild, Input } from '@angular/core';
import * as MindFusion from 'diagram-library';
let Overview = class Overview {
    constructor() {
        this.enabled = true;
        this.allow_zoom = false;
        this.back_color = "#FFFFFF";
        this.scale_factor = 17.825311942959;
        this.auto_s_amount = 0;
        this.scale_mode = MindFusion.Diagramming.ScaleMode.FitAll;
    }
    ngAfterViewInit() {
        if (this.view_id === undefined)
            this.view_id = "overview";
        this.canvas.nativeElement.id = this.view_id;
        var overview = MindFusion.Diagramming.Overview.create(document.getElementById(this.view_id));
        overview.setDiagram(this.diagram);
        overview.setEnabled(this.enabled);
        overview.setAllowZoom(this.allow_zoom);
        overview.setScaleFactor(this.scale_factor);
        overview.setAutoScrollAmount(this.auto_s_amount);
        overview.setScaleMode(this.scale_mode);
        overview.setBackColor(this.back_color);
    }
};
__decorate([
    ViewChild('overview', { static: true })
], Overview.prototype, "canvas", void 0);
__decorate([
    Input('id')
], Overview.prototype, "view_id", void 0);
__decorate([
    Input('diagram')
], Overview.prototype, "diagram", void 0);
__decorate([
    Input('enabled')
], Overview.prototype, "enabled", void 0);
__decorate([
    Input('allowZoom')
], Overview.prototype, "allow_zoom", void 0);
__decorate([
    Input('backColor')
], Overview.prototype, "back_color", void 0);
__decorate([
    Input('scaleFactor')
], Overview.prototype, "scale_factor", void 0);
__decorate([
    Input('autoScrollAmount')
], Overview.prototype, "auto_s_amount", void 0);
__decorate([
    Input('scaleMode')
], Overview.prototype, "scale_mode", void 0);
Overview = __decorate([
    Component({
        selector: 'overview',
        template: `<div style=" width: 200px; height: 200px; ">
                 <canvas #overview width="200" height="200"></canvas>
               </div>`
    })
], Overview);
export { Overview };
