// diagram-library-angular/esm2015/lib/zoomer.component.js
import { __decorate } from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import * as MindFusion from 'diagram-library';
let ZoomControl = class ZoomControl {
    constructor() {
        this.enabled = true;
        this.cornerRadius = 2;
        this.padding = 2;
        this.snapToZoomStep = true;
        this.tick_Position = MindFusion.Controls.TickPosition.Left;
        this.show_label = true;
        this.txt_color = "rgba(77,83,94,1)";
        this.sh_color = "gray";
        this.inn_color = "rgba(91,91,91,1)";
        this.br_color = "rgba(77,83,94,1)";
        this.act_Color = "skyBlue";
        this.fl = "#FFFFFF";
        this.back_color = "#FFFFFF";
        this.scr_step = 10;
        this.z_step = 10;
        this.mx_Z_fac = 200;
        this.mn_Z_fac = 0;
        this.z_fac = 100;
    }
    ngAfterViewInit() {
        if (this.zoom_id === undefined)
            this.zoom_id = "zoomer";
        this.canvas.nativeElement.id = this.zoom_id;
        var zoomer = MindFusion.Controls.ZoomControl.create(document.getElementById(this.zoom_id));
        zoomer.setCornerRadius(this.cornerRadius);
        zoomer.setPadding(this.padding);
        zoomer.setSnapToZoomStep(this.snapToZoomStep);
        zoomer.setTickPosition(this.tick_Position);
        zoomer.setShowLabel(this.show_label);
        zoomer.setTextColor(this.txt_color);
        zoomer.setShadowColor(this.sh_color);
        zoomer.setInnerColor(this.inn_color);
        zoomer.setBorderColor(this.br_color);
        zoomer.setActiveColor(this.act_Color);
        zoomer.setFill(this.fl);
        zoomer.setBackColor(this.back_color);
        zoomer.setScrollStep(this.scr_step);
        zoomer.setZoomStep(this.z_step);
        zoomer.setMaxZoomFactor(this.mx_Z_fac);
        zoomer.setMinZoomFactor(this.mn_Z_fac);
        zoomer.setZoomFactor(this.z_fac);
        zoomer.setEnabled(true);
        zoomer.setTarget(this.diagram);
    }
};
__decorate([
    ViewChild('zoomer', { static: true })
], ZoomControl.prototype, "canvas", void 0);
__decorate([
    Input('id')
], ZoomControl.prototype, "zoom_id", void 0);
__decorate([
    Input('diagram')
], ZoomControl.prototype, "diagram", void 0);
__decorate([
    Input('enabled')
], ZoomControl.prototype, "enabled", void 0);
__decorate([
    Input('cornerRadius')
], ZoomControl.prototype, "cornerRadius", void 0);
__decorate([
    Input('padding')
], ZoomControl.prototype, "padding", void 0);
__decorate([
    Input('snapToZoomStep')
], ZoomControl.prototype, "snapToZoomStep", void 0);
__decorate([
    Input('tickPosition')
], ZoomControl.prototype, "tick_Position", void 0);
__decorate([
    Input('showLabel')
], ZoomControl.prototype, "show_label", void 0);
__decorate([
    Input('textColor')
], ZoomControl.prototype, "txt_color", void 0);
__decorate([
    Input('shadowColor')
], ZoomControl.prototype, "sh_color", void 0);
__decorate([
    Input('innerColor')
], ZoomControl.prototype, "inn_color", void 0);
__decorate([
    Input('borderColor')
], ZoomControl.prototype, "br_color", void 0);
__decorate([
    Input('activeColor')
], ZoomControl.prototype, "act_Color", void 0);
__decorate([
    Input("fill")
], ZoomControl.prototype, "fl", void 0);
__decorate([
    Input('backColor')
], ZoomControl.prototype, "back_color", void 0);
__decorate([
    Input('scrollStep')
], ZoomControl.prototype, "scr_step", void 0);
__decorate([
    Input('zoomStep')
], ZoomControl.prototype, "z_step", void 0);
__decorate([
    Input('maxZoomFactor')
], ZoomControl.prototype, "mx_Z_fac", void 0);
__decorate([
    Input('minZoomFactor')
], ZoomControl.prototype, "mn_Z_fac", void 0);
__decorate([
    Input('zoomFactor')
], ZoomControl.prototype, "z_fac", void 0);
ZoomControl = __decorate([
    Component({
        selector: 'zoom-control',
        template: `<div style="width: 50px; height: 300px;">
                  <canvas #zoomer width="50px" height="300px"></canvas>
               </div>`
    })
], ZoomControl);
export { ZoomControl };
